### PR TITLE
Improve token approval

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Improves the events for token approvals.  As part of addressing these comments https://github.com/cowprotocol/cowswap/pull/655#issuecomment-1149771629

Also, it shows a modal with the error if there's one, or if the user rejects the transaction (there wasn't one).

# To Test

1. Install and activate https://chrome.google.com/webstore/detail/google-analytics-debugger/jnkmfdileelhofjcijamephohjechhna?hl=en 
2. Verify the events for token approvals
3. Verify the console, it should write the event that was sent to analytics

### CODE REVIEW

**Comment by anxolin** [comment]
> But still I see hanging 'Approving' modal when I reject Approve transaction in the connected WC wallet (and no event for this)

I will review the Wallet Connect thing in a separate PR

File: src/custom/pages/Swap/SwapMod.tsx [hunk 1]
```diff
@@ -380,11 +380,47 @@ export default function Swap({
       approveRequired = true
     }
 
-    reportAnalytics('Approve', v2Trade?.inputAmount?.currency.symbol)
     if (approveRequired) {
-      return approveCallback().catch((error) => console.error('Error setting the allowance for token', error))
+      const symbol = v2Trade?.inputAmount?.currency.symbol
+      reportAnalytics('Send Token Approval to Wallet', symbol)
+      return approveCallback()
```
Comment (W3stside) [inline src/custom/pages/Swap/SwapMod.tsx:386]: Why not write this as:
```ts
let swapErrorMessage = 'Sign Token Approval', actionAnalytics = symbol, errorCode
return approveCallback()
    .catch((error) => {
        console.error('Error setting the allowance for token', error)
        if (isRejectRequestProviderError(error)) {
        swapErrorMessage = 'User rejected approving the token'
        actionAnalytics = 'Reject Token Approval'
        } else {
        swapErrorMessage = getProviderErrorMessage(error)
        actionAnalytics = 'Signing Error for Token Approval'
        
        if (error?.code && typeof error.code === 'number') {
          errorCode = error.code
        }
      }

        setSwapState({
            attemptingTxn: false,
            tradeToConfirm,
            showConfirm,
            swapErrorMessage,
            txHash: undefined,
        })
    })
    .finally(() => reportAnalytics(actionAnalytics, symbol, errorCode))
```

**Review by W3stside — APPROVED** [review]

**Comment by elena-zh** [comment]
> > But still I see hanging 'Approving' modal when I reject Approve transaction in the connected WC wallet (and no event for this)
> 
> I will review the Wallet Connect thing in a separate PR

Approved then!

**Review by elena-zh — APPROVED** [review]

File: src/custom/pages/Swap/SwapMod.tsx (hunk 1 above)
Comment (anxolin) [reply src/custom/pages/Swap/SwapMod.tsx:386]: It's a good suggestion, but I would prefer to keep it like this, just to not rely on the mutation of some`swapErrorMessage` outside of the promise

**Comment by nenadV91** [comment]
> Hey @anxolin , everything works perfect with MM wallet!
> 
> I see 'Send' and 'Sign' GA events when connected to WC wallet.
> 
> But still I see hanging 'Approving' modal when I reject Approve transaction in the connected WC wallet (and no event for this) image
> 
> Could you please take a look at it?
> 
> Thanks!

@elena-zh I've tried to reproduce this, but I wasn't able to

**Comment by elena-zh** [comment]
@nenadV91 , I can. 
1. Open https://cowswap.exchange/
2. Press on the Connect wallet button
3. Select WC option
4. Open MM app in a mobile device (both Android and iOS are acceptable)
5. Connect the app to MM wallet in a mobile device
6. Specify From token that is not approved, amount, To token
7. Press on the 'Allow CowSwap to use you token' button
8. Press on the Cancel request in the MM app. 

You will see that 'Approving token' modal is still displayed in the Cowswap app.

**Pull request merged**